bin/offline.cpp

```diff
@@ -11,9 +12,47 @@
 #include <csignal>
 #include <atomic>
 #include <sstream>
+#include <fstream>
 
 using namespace std::literals::chrono_literals;
 
+std::string readFile(const std::string& fileName) {
+    std::ifstream stream(fileName.c_str());
+    if (!stream.good()) {
+        throw std::runtime_error("Cannot read file: " + fileName);
+    }
+    
+    std::stringstream buffer;
+    buffer << stream.rdbuf();
+    stream.close();
+    
+    return buffer.str();
+}
+
+mapbox::geometry::geometry<double> parseGeometry(const std::string& json) {
+    using namespace mapbox::geojson;
+    auto geojson = parse(json);
+    return geojson.match(
+        [](const geometry& geom) {
+            return geom;
+        },
+        [](const feature& feature) {
+            return feature.geometry;
+        },
+        [](const feature_collection& featureCollection) {
+            if (featureCollection.size() < 1) {
+                throw std::runtime_error("No features in feature collection");
+            }
+            geometry_collection geometries;
+            
+            for (auto feature : featureCollection) {
+                geometries.push_back(feature.geometry);
+            }
+            
+            return geometries;
```

Comment:
It would be a recursive definition then. Is there any reason not to support it?

Mainly the types the interface accepts should be implemented right?

Reply:
The streaming Tile cover algorithm we are using works on homogeneous geometry types. A mixed type like `geometry_collection` would have to be run on each of the contained types separately and would return duplicates. It might also not return tiles in a consistent order.